### ISSUE TITLE
Is explicit use of EnumInfo<,,> possible?

### ISSUE THREAD

**Issue by inelisoni** [Tydude4Christ/Enums.NET#23]
Hello! Is it possible to provide an explicit use of EnumInfo<,,> and it's constructor, instead of implicit Activator.CreateInstance?
We use your library in our Unity project, which use AOT compilation and compiler can't determine using of EnumInfo<,,> type. 
In result code doesn't generated for type and we catch exceptions in runtime.

**Comment by TylerBrinkley** [comment]
I'm afraid not. Activator.CreateInstance is used only because there's no alternative. Aren't there ways to explicitly let AOT compilers know of types that shouldn't be tree shook?

**Comment by inelisoni** [comment]
We can only control stripping of the code, but not its generation.
As temporary workaround we add generic overload of GetInfoInfo:
``` c#
internal static IEnumInfo<TEnum> GetEnumInfo<[EnumConstraint] TEnum>() where TEnum : struct
{
    var enumType = typeof(TEnum);
    var underlyingType = Enum.GetUnderlyingType(enumType);

    switch (underlyingType.GetTypeCode())
    {
        case TypeCode.SByte:
            return new EnumInfo<TEnum, SByte, SByteNumericProvider>();
        case TypeCode.Byte:
            return new EnumInfo<TEnum, Byte, ByteNumericProvider>();
        case TypeCode.Int16:
            return new EnumInfo<TEnum, Int16, Int16NumericProvider>();
        case TypeCode.UInt16:
            return new EnumInfo<TEnum, UInt16, UInt16NumericProvider>();
        case TypeCode.Int32:
            return new EnumInfo<TEnum, Int32, Int32NumericProvider>();
        case TypeCode.UInt32:
            return new EnumInfo<TEnum, UInt32, UInt32NumericProvider>();
        case TypeCode.Int64:
            return new EnumInfo<TEnum, Int64, Int64NumericProvider>();
        case TypeCode.UInt64:
            return new EnumInfo<TEnum, UInt64, UInt64NumericProvider>();
        case TypeCode.Boolean:
            return new EnumInfo<TEnum, Boolean, BooleanNumericProvider>();
        case TypeCode.Char:
            return new EnumInfo<TEnum, Char, CharNumericProvider>();
    }
    throw new NotSupportedException($"Enum underlying type of {underlyingType} is not supported");
}
```
and use it for Enums:
``` c#
internal static class Enums<[EnumConstraint] TEnum> where TEnum : struct
{
     public static readonly IEnumInfo<TEnum> Info = Enums.GetEnumInfo<TEnum>();
}
```
UnsafeEnums use non-generic overload of GetEnumInfo.

**Comment by TylerBrinkley** [comment]
I guess it is possible to create it generically but the reason I didn't was for memory consumption considerations. Let me look into this.

**Comment by TylerBrinkley** [comment]
If your code only used NonGenericEnums is there anyway to get that to work?

**Comment by inelisoni** [comment]
Oh, our code use only safe Enums with generic implementation. Sorry, if I misled you.

**Comment by TylerBrinkley** [comment]
It looks like this is possible but I'd rather not do this for now as it does have a code size cost due to generic code explosion with enums. I was looking up Unity development with AOT compilation and it looks like you can use a link.xml file like so to prevent stripping necessary types or methods.

```xml
<linker>
       <assembly fullname="Enums.NET">
               <type fullname="EnumsNET.EnumInfo`3" preserve="all"/>
       </assembly>         
</linker>
```

**Comment by inelisoni** [comment]
In any case thanks for your time.
We tried proposed variant of link.xml:
``` xml
	<assembly fullname="Enums.NET">
		<type fullname="EnumsNET.EnumInfo`3" preserve="all"/>
	</assembly>
```
and even
``` xml
	<assembly fullname="Enums.NET" preserve="all">
		<type fullname="*" preserve="all"/>
	</assembly>
```
but nothing was helpful.
The problem is not code stripping, but code generation. AOT compiler (IL2CPP) generate code for generics during compilation. In case of implicit use, it doesn't generate code. In this case rules in link.xml can be helpful only for Mono scripting backend, but not for IL2CPP.

**Comment by TylerBrinkley** [comment]
@inelisoni Hmm, that makes sense. As there's no workaround except for forking I'll see if I can get this implemented.

**Comment by TylerBrinkley** [comment]
This has been addressed in 2127444.

**Comment by TylerBrinkley** [comment]
@inelisoni FYI, over the weekend I updated the Enums.NET NuGet package to 2.3.2 which incorporated this change. I'd appreciate it if you could validate that it's working for you with Unity AOT compilation as I'm not a Unity developer.
